src/fp/single_word.rs

```diff
@@ -0,0 +1,260 @@
+                let w = self.mu.wrapping_mul(z0);
+                let (r1, r0) = hi_lo((self.p as $W2) * (w as $W2));
+
+                //    z1,z0
+                // +  r1,r0
+                //    =====
+                // cc, z, 0
+                let (_zero, carry) = z0.overflowing_add(r0);
+                let (cc, z) = hi_lo((z1 as $W2) + (r1 as $W2) + (carry as $W2));
+
+                // Final subtraction
+                // If z >= p, then z = z - p
+
+                //    cc, z
+                // -   0, p
+                // ========
+                // b1,s1,s0
+                let (s0, b0) = z.overflowing_sub(self.p);
+                let (_s1, b1) = cc.overflowing_sub(b0 as $W);
+                // if b1 == 1: return z
+                // else:       return s0
+                let mask = (0 as $W).wrapping_sub(b1 as $W);
+                (z & mask) | (s0 & !mask)
+            }
+
+            /// Modular exponentiation, i.e., `x^exp (mod p)` where `p` is
+            /// the modulus. Note that the runtime of this algorithm is
+            /// linear in the bit length of `exp`.
+            pub(crate) fn pow(&self, x: $W, exp: $W) -> $W {
+                let mut t = self.roots[0];
+                for i in (0..<$W>::BITS).rev() {
+                    t = self.mul(t, t);
+                    if (exp >> i) & 1 != 0 {
+                        t = self.mul(t, x);
+                    }
+                }
+                t
+            }
+
+            /// Modular inversion, i.e., x^-1 (mod p) where `p` is the
+            /// modulus. Note that the runtime of this algorithm is
+            /// linear in the bit length of `p`.
+            #[inline(always)]
+            pub(crate) fn inv(&self, x: $W) -> $W {
+                self.pow(x, self.p - 2)
+            }
+
+            /// Negation, i.e., `-x (mod p)` where `p` is the modulus.
+            #[inline(always)]
+            pub(crate) fn neg(&self, x: $W) -> $W {
+                self.sub(0, x)
+            }
+
+            /// Maps an integer to its internal representation. Field
+            /// elements are mapped to the Montgomery domain in order to
+            /// carry out field arithmetic. The result will be in [0, p).
+            #[inline(always)]
+            pub(crate) fn montgomery(&self, x: $W) -> $W {
+                self.modp(self.mul(x, self.r2))
+            }
+
+            /// Maps a field element to its representation as an integer.
+            /// The result will be in [0, p).
+            #[inline(always)]
+            pub(crate) fn residue(&self, x: $W) -> $W {
+                self.modp(self.mul(x, 1))
+            }
+
+            #[inline(always)]
+            fn modp(&self, x: $W) -> $W {
+                let (z, carry) = x.overflowing_sub(self.p);
+                let m = (0 as $W).wrapping_sub(carry as $W);
+                z.wrapping_add(m & self.p)
+            }
+        }
+
+        #[cfg(test)]
```

Comment:
Instead of making this code match the API of the tests, I'd suggest making the tests generic in the word size.

Reply:
I have updated the code and tests to rely on a generic implementation.